android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/ButtonFactory.java

```diff
@@ -0,0 +1,123 @@
+                                       CommonListener listener) throws Exception {
+        String openMrsEntityParent = jsonObject.getString("openmrs_entity_parent");
+        String openMrsEntity = jsonObject.getString("openmrs_entity");
+        String openMrsEntityId = jsonObject.getString("openmrs_entity_id");
+        String relevance = jsonObject.optString("relevance");
+
+        List<View> views = new ArrayList<>(1);
+        JSONArray canvasIds = new JSONArray();
+        Button button = new Button(context);
+        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
+                ViewGroup.LayoutParams.MATCH_PARENT,
+                ViewGroup.LayoutParams.WRAP_CONTENT);
+        layoutParams.bottomMargin = context.getResources()
+                .getDimensionPixelSize(R.dimen.extra_bottom_margin);
+        button.setLayoutParams(layoutParams);
+        button.setTextSize(TypedValue.COMPLEX_UNIT_PX, context.getResources()
+                .getDimension(R.dimen.button_text_size));
+        button.setTextColor(context.getResources().getColor(android.R.color.white));
+        button.setHeight(context.getResources().getDimensionPixelSize(R.dimen.button_height));
+
+        String hint = jsonObject.optString("hint");
+        if (!TextUtils.isEmpty(hint)) {
+            button.setText(hint);
+        }
+
+        button.setId(ViewUtil.generateViewId());
+        canvasIds.put(button.getId());
+
+        button.setTag(R.id.key, jsonObject.getString("key"));
+        button.setTag(R.id.openmrs_entity_parent, openMrsEntityParent);
+        button.setTag(R.id.openmrs_entity, openMrsEntity);
+        button.setTag(R.id.openmrs_entity_id, openMrsEntityId);
+        button.setTag(R.id.type, jsonObject.getString("type"));
+        button.setTag(R.id.address, stepName + ":" + jsonObject.getString("key"));
+
+        if (jsonObject.has("read_only")) {
+            button.setEnabled(!jsonObject.getBoolean("read_only"));
+            button.setFocusable(!jsonObject.getBoolean("read_only"));
+        }
+
+        JSONObject action = jsonObject.optJSONObject("action");
+        if (action != null) {
+            jsonObject.put("value", "false");
+            jsonObject.getJSONObject("action").put("result", false);
+            final String behaviour = action.optString("behaviour");
+            button.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    try {
+                        String addressString = (String) v.getTag(R.id.address);
+                        if (!TextUtils.isEmpty(addressString)) {
+                            String[] address = addressString.split(":");
+                            JSONObject jsonObject = ((JsonApi) context)
+                                    .getObjectUsingAddress(address);
+                            jsonObject.getJSONObject("action").put("result", false);
```

Comment:
is there a reason to set this here and not after the switch, is something happening before the var is updated again in the switch

If not,

```
bool actionResult = false;
...
    case BEH...:
         actionResult = true;
...
jsonObject.getJSONObject("action").put("result", actionResult);
```
 
declare var here, replace lines 95 and 100 with var assignment, move this line to after switch and call with var